### ISSUE TITLE
Possible issue - SourceNode circular dependencies

### ISSUE THREAD

**Issue by micimize** [gql-dart/gql#80]
I like the `SourceNode` approach to import handling, but won't this result in an infinite loop in the event of a circular import?

https://github.com/gql-dart/gql/blob/e8ec7e341099e9916b14c72c5f39d41d7747c699/gql_build/lib/src/utils/reader.dart#L63-L70

**Comment by klavs** [comment]
@micimize, you are right! I just assumed that in practice it shouldn't happen. It would be better if I handled it, though.

At the time I wanted to get this working as a PoC, so I dropped the code you already wrote. I just realized at the time that there's a second dimension to the problem — all imports must be resolved to a cannonical path so that an import appearing more than once is read only once, but still has multiple nodes in the SourceNode tree.

If you are up to it, feel free to PR the fix.
